lib/mli_parser.ml

```diff
@@ -1,207 +1,128 @@
-        as e ->
-          acc [ e ]
-      | { value = `Tag tag; _ } -> (
-          match tag with
-          | `Deprecated blocks -> acc blocks
-          | `Param (_, blocks) -> acc blocks
-          | `Raise (_, blocks) -> acc blocks
-          | `Return blocks -> acc blocks
-          | `See (_, _, blocks) -> acc blocks
-          | `Before (_, blocks) -> acc blocks
-          | _ -> [])
-      | { value = `Heading _; _ } -> [])
-    (Odoc_parser.ast parsed)
-  |> List.concat
 
-let docstrings lexbuf =
+  List.fold_left fold_fn acc (ast parsed)
+
+(* This function handles string containing ocaml code. It parses it as ocaml
+   via compiler-libs, then for each odoc-formatted comment it then parses
+   that via odoc-parser. The end result is a list of metadata about the code
+   blocks within the comments. The result is given as an in-order list of
+   [Code_block.t] values. *)
+let docstring_code_blocks str =
+  Lexer.handle_docstrings := true;
+  Lexer.init ();
+  let lexbuf = Lexing.from_string str in
   let rec loop list =
     match Lexer.token_with_comments lexbuf with
     | Parser.EOF -> list
     | Parser.DOCSTRING docstring ->
-        let docstring =
-          ( Docstrings.docstring_body docstring,
-            Docstrings.docstring_loc docstring )
+        let body = Docstrings.docstring_body docstring in
+        let loc = Docstrings.docstring_loc docstring in
+
+        (* odoc-parser adjusts for the initial [** *)
+        let adjustment = 3 (* String.length "(**" *) in
+
+        let location =
+          { loc.loc_start with pos_cnum = loc.loc_start.pos_cnum + adjustment }
         in
-        loop (docstring :: list)
+        loop (extract_code_block_info list ~location ~docstring:body)
     | _ -> loop list
   in
   loop [] |> List.rev
 
-let convert_pos (p : Lexing.position) (pt : Odoc_parser.Loc.point) =
-  { p with pos_lnum = pt.line; pos_cnum = pt.column }
-
-let convert_loc (loc : Location.t) (sp : Odoc_parser.Loc.span) =
-  let loc_start = convert_pos loc.loc_start sp.start in
-  let loc_end = convert_pos loc.loc_end sp.end_ in
-  { loc with loc_start; loc_end }
-
-let docstring_code_blocks str =
-  Lexer.handle_docstrings := true;
-  Lexer.init ();
-  List.map
-    (fun (docstring, (cmt_loc : Location.t)) ->
-      let location =
-        {
-          cmt_loc.loc_start with
-          pos_cnum = cmt_loc.loc_start.pos_cnum + docstring_start_adjustment;
-        }
-      in
-      let blocks = extract_code_blocks ~location ~docstring in
-      List.map
-        (fun (b : Code_block.t) -> (b, convert_loc cmt_loc b.location))
-        blocks)
-    (docstrings (Lexing.from_string str))
-  |> List.concat
-
-let make_block ~loc code_block =
+(* Given code block metadata and the original file, this function splices the
+   contents of the code block from the original text and creates an Mdx
+   Block.t, or reports the error (e.g., from invalid tags) *)
+let make_block code_block orig =
```

Comment:
I think `orig` is a poor name, maybe `let extract_block code_block contents`?

Reply:
Agreed! I'll change it :-)

Reply:
I think `file_contents` is better - that's actually the name of the variable passed to this function below.